<a href="https://colab.research.google.com/github/Ehizson/HDSC-S2023-Projects/blob/main/Stage_B_Project.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## **TAG ALONG PROJECT ON REGRESSION IN MACHINE LEARNING**

## **Michael Godwin Ehizojie**

### **ID: 1603dc278b81f000**

In [1]:
#import libraries

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

from scipy import stats
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.linear_model import Ridge
from sklearn.linear_model import Lasso
from sklearn.metrics import r2_score
from sklearn.preprocessing import MinMaxScaler
from sklearn.metrics import mean_absolute_error
from sklearn.metrics import mean_squared_error

**Data Preparation**

In [2]:
#loading dataset

Energy_data = pd.read_csv('/content/energydata_complete.csv')
Energy_data.head()

date  Appliances  lights     T1       RH_1    T2       RH_2  \
0  2016-01-11 17:00:00          60      30  19.89  47.596667  19.2  44.790000   
1  2016-01-11 17:10:00          60      30  19.89  46.693333  19.2  44.722500   
2  2016-01-11 17:20:00          50      30  19.89  46.300000  19.2  44.626667   
3  2016-01-11 17:30:00          50      40  19.89  46.066667  19.2  44.590000   
4  2016-01-11 17:40:00          60      40  19.89  46.333333  19.2  44.530000   

      T3       RH_3         T4  ...         T9   RH_9     T_out  Press_mm_hg  \
0  19.79  44.730000  19.000000  ...  17.033333  45.53  6.600000        733.5   
1  19.79  44.790000  19.000000  ...  17.066667  45.56  6.483333        733.6   
2  19.79  44.933333  18.926667  ...  17.000000  45.50  6.366667        733.7   
3  19.79  45.000000  18.890000  ...  17.000000  45.40  6.250000        733.8   
4  19.79  45.000000  18.890000  ...  17.000000  45.40  6.133333        733.9   

   RH_out  Windspeed  Visibility  Tdewpoint        rv1        rv2  
0    92.0   7.000000   63.000000        5.3  13.275433  13.275433  
1    92.0   6.666667   59.166667        5.2  18.606195  18.606195  
2    92.0   6.333333   55.333333        5.1  28.642668  28.642668  
3    92.0   6.000000   51.500000        5.0  45.410389  45.410389  
4    92.0   5.666667   47.666667        4.9  10.084097  10.084097  

[5 rows x 29 columns]

In [3]:
Energy_data.shape

(19735, 29)

In [4]:
#checking for missing numbers

Energy_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 19735 entries, 0 to 19734
Data columns (total 29 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   date         19735 non-null  object 
 1   Appliances   19735 non-null  int64  
 2   lights       19735 non-null  int64  
 3   T1           19735 non-null  float64
 4   RH_1         19735 non-null  float64
 5   T2           19735 non-null  float64
 6   RH_2         19735 non-null  float64
 7   T3           19735 non-null  float64
 8   RH_3         19735 non-null  float64
 9   T4           19735 non-null  float64
 10  RH_4         19735 non-null  float64
 11  T5           19735 non-null  float64
 12  RH_5         19735 non-null  float64
 13  T6           19735 non-null  float64
 14  RH_6         19735 non-null  float64
 15  T7           19735 non-null  float64
 16  RH_7         19735 non-null  float64
 17  T8           19735 non-null  float64
 18  RH_8         19735 non-null  float64
 19  T9  

In [5]:
#duplicate checking

Energy_data.duplicated().any()

False

In [7]:
#statistical summary

Energy_data.describe()

Appliances        lights            T1          RH_1            T2  \
count  19735.000000  19735.000000  19735.000000  19735.000000  19735.000000   
mean      97.694958      3.801875     21.686571     40.259739     20.341219   
std      102.524891      7.935988      1.606066      3.979299      2.192974   
min       10.000000      0.000000     16.790000     27.023333     16.100000   
25%       50.000000      0.000000     20.760000     37.333333     18.790000   
50%       60.000000      0.000000     21.600000     39.656667     20.000000   
75%      100.000000      0.000000     22.600000     43.066667     21.500000   
max     1080.000000     70.000000     26.260000     63.360000     29.856667   

               RH_2            T3          RH_3            T4          RH_4  \
count  19735.000000  19735.000000  19735.000000  19735.000000  19735.000000   
mean      40.420420     22.267611     39.242500     20.855335     39.026904   
std        4.069813      2.006111      3.254576      2.042884      4.341321   
min       20.463333     17.200000     28.766667     15.100000     27.660000   
25%       37.900000     20.790000     36.900000     19.530000     35.530000   
50%       40.500000     22.100000     38.530000     20.666667     38.400000   
75%       43.260000     23.290000     41.760000     22.100000     42.156667   
max       56.026667     29.236000     50.163333     26.200000     51.090000   

       ...            T9          RH_9         T_out   Press_mm_hg  \
count  ...  19735.000000  19735.000000  19735.000000  19735.000000   
mean   ...     19.485828     41.552401      7.411665    755.522602   
std    ...      2.014712      4.151497      5.317409      7.399441   
min    ...     14.890000     29.166667     -5.000000    729.300000   
25%    ...     18.000000     38.500000      3.666667    750.933333   
50%    ...     19.390000     40.900000      6.916667    756.100000   
75%    ...     20.600000     44.338095     10.408333    760.933333   
max    ...     24.500000     53.326667     26.100000    772.300000   

             RH_out     Windspeed    Visibility     Tdewpoint           rv1  \
count  19735.000000  19735.000000  19735.000000  19735.000000  19735.000000   
mean      79.750418      4.039752     38.330834      3.760707     24.988033   
std       14.901088      2.451221     11.794719      4.194648     14.496634   
min       24.000000      0.000000      1.000000     -6.600000      0.005322   
25%       70.333333      2.000000     29.000000      0.900000     12.497889   
50%       83.666667      3.666667     40.000000      3.433333     24.897653   
75%       91.666667      5.500000     40.000000      6.566667     37.583769   
max      100.000000     14.000000     66.000000     15.500000     49.996530   

                rv2  
count  19735.000000  
mean      24.988033  
std       14.496634  
min        0.005322  
25%       12.497889  
50%       24.897653  
75%       37.583769  
max       49.996530  

[8 rows x 28 columns]

**Variables Description**

Appliances : It is our target variable ranging from 10 to 1080.Had some rare peak values.

lights : It has mostly zero values and it has some outliers.

(All the temperature measures here are measured in Celsius)

T1(Temperature in Kitchen Area) : These temperature looks descent without any outliers and seems like the data doesn't have cold climate in any temperature features becuase the min value is more than 0 for all the features.

T2(Temperature in Living Room) : These temperature looks descent without any outliers.

RH1 to RH2 (Room Humidity): It is the measure of humidity in different rooms in the house.

Windspeed : This Windspeed has some outliers.

From Chievers weather station we have some weather date like Hg RHout, Visibility, Tdewpoint etc.

In [8]:
#12

X = Energy_data.drop('T6', axis=1)
y = Energy_data['T6']

In [9]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.3, random_state = 42)

print("X_train : ",X_train.shape)
print("y_train : ",y_train.shape)
print("X_test : ",X_test.shape)
print("y_test : ",y_test.shape)

X_train :  (13814, 28)
y_train :  (13814,)
X_test :  (5921, 28)
y_test :  (5921,)


In [11]:
model = LinearRegression()
model.fit(X_train[['T2']], y_train)
prediction = model.predict(X_test[['T2']])

print(round(r2_score(y_test, prediction), 2))

0.64


In [12]:
#13

copy = Energy_data.copy()

In [13]:
copy = copy.drop(['date', 'lights'], axis=1)
min_max = MinMaxScaler()
scaled_df = min_max.fit_transform(copy)
normalized_df = pd.DataFrame(scaled_df, columns = copy.columns)

In [14]:
x = normalized_df.drop('Appliances', axis=1)
y = normalized_df['Appliances']

In [15]:
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size = 0.3, random_state = 42)

print("x_train : ",x_train.shape)
print("y_train : ",y_train.shape)
print("x_test : ",x_test.shape)
print("y_test : ",y_test.shape)

x_train :  (13814, 26)
y_train :  (13814,)
x_test :  (5921, 26)
y_test :  (5921,)


In [16]:
model2 = LinearRegression()
model2.fit(x_train, y_train)
prediction = model2.predict(x_test)
ab_error = round(mean_absolute_error(y_test, prediction), 2)
ab_error

0.05

In [18]:
#14

residual = round(np.sum(np.square(prediction - y_test)), 2)
residual

45.35

In [19]:
#15

root_mse = round(mean_squared_error(y_test, prediction, squared=False), 3)
root_mse

0.088

In [20]:
#16

coefficient = round(r2_score(y_test, prediction), 2)
coefficient

0.15

In [22]:
#17

def feature_weights (model_name, feat, col_name):
    weights = pd.Series(model_name.coef_, feat.columns).sort_values()
    df_weights = pd.DataFrame(weights).reset_index()
    df_weights.columns = ["Features", col_name]
    df_weights[col_name].round(2)
    return df_weights

Regression_model_weights = feature_weights(model2, x_train, "Regression_model_weights")
Regression_model_weights

Features  Regression_model_weights
0          RH_2                 -0.456698
1         T_out                 -0.321860
2            T2                 -0.236178
3            T9                 -0.189941
4          RH_8                 -0.157595
5        RH_out                 -0.077671
6          RH_7                 -0.044614
7          RH_9                 -0.039800
8            T5                 -0.015657
9            T1                 -0.003281
10          rv1                  0.000770
11          rv2                  0.000770
12  Press_mm_hg                  0.006839
13           T7                  0.010319
14   Visibility                  0.012307
15         RH_5                  0.016006
16         RH_4                  0.026386
17           T4                  0.028981
18    Windspeed                  0.029183
19         RH_6                  0.038049
20         RH_3                  0.096048
21           T8                  0.101995
22    Tdewpoint                  0.117758
23           T6                  0.236425
24           T3                  0.290627
25         RH_1                  0.553547

In [23]:
#18

ridge_reg = Ridge(alpha=0.4)
ridge_reg.fit(x_train, y_train)
pred = ridge_reg.predict(x_test)
ridge_error = round(mean_squared_error(y_test, pred, squared=False), 3)
ridge_error

0.088

In [24]:
#19

lasso_reg = Lasso(alpha=0.001)
lasso_reg.fit(x_train, y_train)

weights = feature_weights(lasso_reg, x_train, "Weight")
weights

Features    Weight
0        RH_out -0.049557
1          RH_8 -0.000110
2            T1  0.000000
3     Tdewpoint  0.000000
4    Visibility  0.000000
5   Press_mm_hg -0.000000
6         T_out  0.000000
7          RH_9 -0.000000
8            T9 -0.000000
9            T8  0.000000
10         RH_7 -0.000000
11          rv1 -0.000000
12           T7 -0.000000
13           T6  0.000000
14         RH_5  0.000000
15           T5 -0.000000
16         RH_4  0.000000
17           T4 -0.000000
18         RH_3  0.000000
19           T3  0.000000
20         RH_2 -0.000000
21           T2  0.000000
22         RH_6 -0.000000
23          rv2 -0.000000
24    Windspeed  0.002912
25         RH_1  0.017880

In [25]:
#20

prediction = lasso_reg.predict(x_test)
rms_error = round(mean_squared_error(y_test, prediction, squared=False), 3)
rms_error

0.094